### ISSUE TITLE
Slow rescheduling of all cards

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs4anki-helper#37]
Rescheduling all cards (not only the cards studied today) with the new version is noticeably slower than with the old version.

_Originally posted by @user1823 in https://github.com/open-spaced-repetition/fsrs4anki-helper/issues/25#issuecomment-1445370690_

I feel the same today when rescheduling all the cards as a part of testing. I used to reschedule all my cards everyday before https://github.com/open-spaced-repetition/fsrs4anki-helper/issues/25 was resolved. It never took so much time to reschedule the cards as it takes now.

**Comment by L-M-Sherlock** [comment]
But this PR doesn't modify the main process of rescheduling. I don't know which part induces the slowing down. And which old version did you use? Maybe it is a matter of some earlier PRs.

**Comment by user1823** [comment]
The old version I am referring to is the version on AnkiWeb just before 16100ad955c0289affffe3d72d3090ea2ffba05d was merged.

I am not a developer. So, I can't point out why it induces the slow down. But I definitely feel it.

If you can't find out the cause, I suggest keeping this issue open so that other people can confirm and suggest solutions.

Also, I would suggest sharing here the version of the helper just before the above commit was merged so that other people can compare the time taken to reschedule.

**Comment by L-M-Sherlock** [comment]
The older version: [fsrs4anki-helper.zip](https://github.com/open-spaced-repetition/fsrs4anki-helper/files/10891968/fsrs4anki-helper.zip)

**Comment by cjdduarte** [comment]
@user1823 how many cards do you have?
I have about 1,000 studied and I didn't notice any difference in performance.

**Comment by user1823** [comment]
I have about 8000 cards.

**Comment by cjdduarte** [comment]
I did a stress test here on a deck with **35,000** studied cards.

The reason for the slowness is the "load balance" function

With it disabled, the recalculation of 35 thousand cards took 11 seconds.

With it enabled, it took about 1 minute and 20 seconds.

I downloaded this old version and testing again, it took 1 minute and 48 seconds.

Did you use the "load balance" version before?

**Comment by user1823** [comment]
I did a test on my deck again with the older version posted above.

- With load balance enabled, rescheduling took 5 minutes 46 seconds.
- With load balance disabled, rescheduling took 10 seconds.

With the latest version from AnkiWeb,

- With load balance enabled, rescheduling took 8 minutes 20 seconds.
- With load balance disabled, rescheduling took 10 seconds.

So, it is clear that load balancing is the main culprit. But, still rescheduling with the new version took me about 2.5 minutes more than the old version.

For anyone who is interested in trying it with my deck, here it is: [Default.zip](https://github.com/open-spaced-repetition/fsrs4anki-helper/files/10892239/Default.zip)

Before using, rename it to change `.zip` to `.apkg`

**Comment by cjdduarte** [comment]
I did a test with **your deck.**

With the latest version from AnkiWeb,

- With load balance enabled, rescheduling took 29 seconds.
- With load balance disabled, rescheduling took 3 seconds.

My computer is a Ryzen 5600G, so it processed faster.

But about the slowness, I imagine that a longer time is acceptable because you need to evaluate the number of cards per "ivl" interval.

Maybe the code can be optimized. I noticed a bug in this function:
`due_cards = mw.col.db.scalar("select count() from cards where due = ? and queue = 2", check_due)`

Wouldn't it be a case of throwing the values into a list and working with the list?
I did a simulation in chatgpt.

```
import sqlite3

# Connect to the database
conn = sqlite3.connect('mydatabase.db')
cursor = conn.cursor()

# Get the minimum and maximum value of due
min_due = cursor.execute('SELECT MIN(due) FROM cards').fetchone()[0]
max_due = cursor.execute('SELECT MAX(due) FROM cards').fetchone()[0]

# Create a list with the total varying from the minimum value of due to the last value of due
totals = []
for due in range(min_due, max_due + 1):
    total = cursor.execute('SELECT COUNT(*) FROM cards WHERE due = ?', (due,)).fetchone()[0]
    totals.append(total)

# Close the connection to the database
conn.close()

# Print the list of totals
print(totals)

```

I wouldn't be able to tell if there are loops or database accesses more than usual.
**For me the code and time works perfectly, but I believe that something can be optimized.**

**Comment by L-M-Sherlock** [comment]
> # Create a list with the total varying from the minimum value of due to the last value of due
> totals = []
> for due in range(min_due, max_due + 1):
>     total = cursor.execute('SELECT COUNT(*) FROM cards WHERE due = ?', (due,)).fetchone()[0]
>     totals.append(total)

It could be faster than before If we only calculate the number of due cards per day once. However, the number varies during the rescheduling.

**Comment by L-M-Sherlock** [comment]
I found that counting rated cards also slowed down the rescheduling because it made an extra query from the database. I will optimize it in the next few versions.

**Comment by cjdduarte** [comment]
> > # Create a list with the total varying from the minimum value of due to the last value of due
> > totals = []
> > for due in range(min_due, max_due + 1):
> > total = cursor.execute('SELECT COUNT(*) FROM cards WHERE due = ?', (due,)).fetchone()[0]
> > totals.append(total)
> 
> It could be faster than before If we only calculate the number of due cards per day once. However, the number varies during the rescheduling.

Here was a suggestion. But I don't know if it's the best solution.
It is important to reinforce that it is currently working as it should, but slower.

**Comment by user1823** [comment]
It tested it on the same deck as above (in a test profile).
- With load balance enabled, rescheduling took 31 seconds.
- With load balance disabled, rescheduling took 9 seconds.

So, it is not only 7 times faster than before, but a whopping 16 times faster. I am truly amazed that how such a small change in code can result in such a great boost in performance.

With this change, I believe that the issue is resolved, and frankly, the performance greatly exceeds my expectations. However, if you have further optimizations in mind, please do them as well. Otherwise, this issue can be closed.

**Comment by user1823** [comment]
One question: Does this change lead to any loss in functionality? For example, does it still calculate the number of due cards per day during rescheduling or it calculates it only once?

**Comment by L-M-Sherlock** [comment]
> Does this change lead to any loss in functionality? For example, does it still calculate the number of due cards per day during rescheduling or it calculates it only once?

It still tracks the amount of due cards per day during rescheduling. So it would behave as usual. If you meet some unexpected behaviors from the new code, please report it here.

**Comment by cjdduarte** [comment]
I did some more exhaustive tests and it's OK for me.
Impressed with the performance improvement. What has changed?

**Comment by cjdduarte** [comment]
@L-M-Sherlock I will open another issue to confirm a suspicion in the tests I did here.
but it is another matter.
